### DISCUSSION TITLE
Version 5 Roadmap

### DISCUSSION THREAD

**Discussion by aivascu** [AutoFixture/AutoFixture#1327] [General]
Hello everyone,

 **AutoFixture v5** development is currently in progress and first releases have been published, on the ~[MyGet feed](https://www.myget.org/gallery/autofixture)~ [NuGet feed](https://www.nuget.org/packages/AutoFixture). This thread will collect information on the planned and implemented features as well as deprecated features in the upcoming version.

[![Nuget (with prereleases)](https://img.shields.io/nuget/vpre/AutoFixture)](https://www.nuget.org/packages/AutoFixture)

If you have any ideas please share them, in the [Discussions](https://github.com/AutoFixture/AutoFixture/discussions/categories/ideas).

_The list below is incomplete and subject to change._

### AutoFixture

- [x] ~Add support for .NET Standard 2.1~ #1324
- [x] Add support for `IAsyncEnumerable<>` #1325
- [x] Add NET 5 target to all projects #1443 
- [x] Add NET 6 target to all projects #1453 
- [x] Add support for `IAsyncEnumerator<>`
- [ ] Add support for record types
  - [ ] Add support for `init` accessors
  - [ ] Add postprocessor API for constructor parameters
- [ ] Improve support for value objects

### AutoFixture.NUnit3

- [ ] Allow `[Frozen]` to be used for injected test parameters 
- [ ] Add `[MemberAutoData]`
- [ ] Add `[ClassAutoData]`

### AutoFixture.NUnit4

- [x] Migrate existing library
- [ ] Update `AutoData` attibutes (see NUnit 3 changes above)

### AutoFixture.XUnit2

- [x] Allow `[Frozen]` to be used for injected test parameters #1433 
- [x] Fix `[MemberAutoData]` #1433
- [x] Add `[ClassAutoData]` #1433 

### AutoFixture.NUnit2

- [x] Enable running tests using `dotnet test`

### AutoFixture.AutoMoq

- [x] Raise minimum supported Moq version to `v4.7.49` ~for `netstandard1.5` or greater~

### AutoFixture.Idioms

- [x] Fix `EmptyGuidBehaviorExpectation` not throwing on `ArgumentException`
- [ ] Skip nullable parameters in `GuardClauseAssertion` #1328
- [ ] Add assertions for equality operators

### AutoFixture.*

- [x] Drop support for .NET Core 1.x
- [x] Drop support for .NET Standard 1.x
- [x] Raise minimum .NET Framework version to v4.6.1
- [x] Raised minimum supported version of `NUnit` to `3.7.0` #1443 
- [x] Raised minimum supported version of `xunit.extensibility.core` to `2.2.0` #1433
- [x] Raise minimum supported version of `FakeItEasy` to `6.0.0` #1429
- [x] Raise minimum supported version of `Moq` to `4.7.49` #1429
- [x] Raise minimum supported version of `Foq` to `1.8.0` #1429
- [x] Raise minimum supported version of `FsCheck` to `2.14.0`  #1429
- [x] Raise minimum supported version of `NSubstitute` to ~`3.0.0`~ `4.0.0`  #1429
- [x] Raise minimum supported version of `System.ComponentModel.Annotations` to `4.4.0` #1429
- [x] Raise minimum supported version of `FSharp.Core` to `4.2.3` #1429
- [x] Replace .NET Framework assembly reference to `System.ComponentModel.DataAnnotations` with package reference #1429
- [x] Drop support for AutoFixture.RhinoMock #1502 
- [x] Drop support for AutoFixture.NUnit2 #1502 
- [x] Drop support for AutoFixture.XUnit1 #1502 
- [ ] Hide/Remove `[Obsolete]` API marked as error
- [ ] Mark as error, non-error `[Obsolete]` API

**Comment by rcdailey** [comment]
Does supporting `record` types imply support for init-only properties? At the moment there's no good mechanism to initialize a record type's init-only properties. This probably stems from the existing struggles with proper constructor support, especially with the use of a mocking framework via `AutoData`.

**Reply by aivascu** [reply]
@rcdailey actually I am planning to do both, support for `init` accessors and an improved API for constructor parameters. However if the implementation will drag for too long I might move it to `v6`.

**Comment by Kralizek** [comment]
Should we also consider updating:
- up .net framework support to version to 4.8 (currently [we support 4.6.1 which has reached EOL](https://docs.microsoft.com/en-us/lifecycle/products/microsoft-net-framework)
- up current .net sdk to 6.0.100 with rolling forward for minor version (currently in v5: 5.0.401 with no roll forward configured)


```json
{
  "sdk": {
    "version": "6.0.100",
    "rollForward": "latestMinor"
  }
}
```

**Comment by eduarrodrigueztb** [comment]
Hi Thank you for this awesome library, Is the AutoFixture library still in development? I see this V5 comments are from 1 year ago so not sure if the development somehow stopped ?

**Reply by isberg** [reply]
@eduarrodrigueztb there definitely seems to be progress towards V5. A new pre-release version was made available on nuget this morning. I believe heroic efforts are made on very limited time.

**Reply by halling** [reply]
@aivascu Is the AutoFixture library still in development?

**Reply by aivascu** [reply]
@halling I have opened a new thread to answer this question #1469.
